Added enum export to documentation generation

penev92 opened the pull request:
The last* (planned) step of the documentation rework for this release cycle, it follows up on https://github.com/OpenRA/OpenRA/pull/19948, https://github.com/OpenRA/OpenRA/pull/19986, https://github.com/OpenRA/OpenRA/pull/20087, https://github.com/OpenRA/OpenRA/pull/20250 and https://github.com/OpenRA/OpenRA/pull/20268 by listing relevant enum types and their possible values. By "relevant" I mean those being referenced by traits/weapons/sequences as a field type.

*Depending on whether and how much time allows, I may continue with other things like `ISpriteSequenceLoader`s (those in particular should be trivial).

Reviewing with whitespace diff disabled is recommended.

P.S.: I have updated the `Dev` versions of the trait/weapon/sequence docs on the test version of the website:  https://docs.openra.net/en/test/

penev92 committed:
Added enum export to documentation generation

Mailaender commented on packaging/format-docs.py:
```diff
@@ -81,14 +93,21 @@ def format_docs(version, collectionName, types):
+                        print(f'| {prop["PropertyName"]} | {prop["DefaultValue"] or ""} | {typeName} | {prop["Description"]} |')
+
+    if len(relatedEnums) > 0:
+        print('\n# Related enums:\n')
```
```suggestion
        print('\n## Possible Values\n')
```

maybe? To avoid programmer speak.

penev92 replied:
I will switch this to "Related value types:" to catch enums and potentially future complex types if we ever agree on my POCO types for traitInfo fields idea.
Also there is a suggestion below that I add "Possible values:" in front of the enum's values, so those two wouldn't go well together.

PunkPun commented on OpenRA.Mods.Common/UtilityCommands/ExtractSequenceDocsCommand.cs:
```diff
@@ -48,7 +48,10 @@ void IUtilityCommand.Run(Utility utility, string[] args)
+
+			var sequenceTypesInfo = sequenceTypes
+				.Where(x => !x.ContainsGenericParameters && !x.IsAbstract)
 				.Where(x => x.Name != nameof(FileNotFoundSequence)) // NOTE: This is the simplest way to exclude FileNotFoundSequence, which shouldn't be added.
```
this should really be reusing the .Where

penev92 replied:
I think this is much better for readability and for the comment that follows though.

PunkPun replied:
it can stay in new line, won't make it any less readable

PunkPun commented on packaging/format-docs.py:
```diff
@@ -70,6 +75,13 @@ def format_docs(version, collectionName, types):
+                    # in which case get a link to the enum's definition.
+                    typeName = prop["UserFriendlyType"]
+                    if prop["InternalType"] in enumNames:
+                        typeName = format_type_name(prop["InternalType"], True)
```
Perhaps move the word `enum` to here? "Enum: PlayerRelationships"

<img width="502" alt="Screenshot 2022-09-12 at 19 50 28" src="https://user-images.githubusercontent.com/37534529/189711379-e10e30a8-f2da-48e2-869e-5eb16420b333.png">

penev92 replied:
I don't think that goes too great with the other requested changes. Let's see how they go first.

PunkPun commented on packaging/format-docs.py:
```diff
@@ -81,14 +93,21 @@ def format_docs(version, collectionName, types):
+        print('\n# Related enums:\n')
+        for relatedEnum in relatedEnums:
+            values = [value['Value'] for value in relatedEnum["Values"]]
+            print(f"#### {relatedEnum['Name']}")
```
```suggestion
            print(f"### {relatedEnum['Name']}")
```

To match trait names

PunkPun commented on packaging/format-docs.py:
```diff
@@ -81,14 +93,21 @@ def format_docs(version, collectionName, types):
+        for relatedEnum in relatedEnums:
+            values = [value['Value'] for value in relatedEnum["Values"]]
+            print(f"#### {relatedEnum['Name']}")
+            print(f"{', '.join(values)}\n")
```
```suggestion
            print(f"Possible values: {', '.join(values)}\n")
```

Something this this should be added, can't leave the values bare

penev92 commented:
Updated both the PR and the test docs site.

PunkPun commented on packaging/format-docs.py:
```diff
@@ -46,6 +50,7 @@ def format_docs(version, collectionName, types):
 				"Please do not edit it directly, but instead add new `[Desc(\"String\")]` tags to the source code.\n")
 
     print(f"Listed below are {explanation}.")
+    print(f"Related enums with their possible values are listed [at the bottom](#related-enums).")
```
This link doesn't work

PunkPun commented on packaging/format-docs.py:
```diff
@@ -81,14 +93,21 @@ def format_docs(version, collectionName, types):
+                        print(f'| {prop["PropertyName"]} | {prop["DefaultValue"] or ""} | {typeName} | {prop["Description"]} |')
+
+    if len(relatedEnums) > 0:
+        print('\n# Related value types:\n')
```
```suggestion
        print('\n## Related value types:\n')
```

To match namespace formatting. Nothing uses just one hash so it looks very out of place

penev92 replied:
I was rather thinking we could add `# Weapons`/`# Traits`/`# SpriteSequences` above those to match this one. 🤔

PunkPun replied:
Perhaps, but it would still look out of place

Mailaender reviewed (dismissed):
Looking good.

penev92 commented:
Changed the "Related value types:" title to "Related value types (enums):" and added a matching "Traits"/"Weapons"/"Sprite Sequences" title at the top. Let's see how that looks (I already updated the docs site. Might need a force-refresh).
Also fixed the top link.

penev92 committed:
Added referencing types to enum definitions

penev92 committed:
Added some sorting to documentation output

penev92 commented:
Added a couple of things as requested on Discord:
 - Added a list of referencing types to each enum definition
 - Sorted some generated lists (traits inside namespaces, referencing types for enums and required traits for traits)

P.S.: Also fixed the SpriteSequences output Markdown file's name and the utility command.

penev92 committed:
Renamed SpriteSequence documentation command and output file

PunkPun commented:
This looks rather jarring

<img width="1096" alt="Screenshot 2022-09-14 at 10 52 47" src="https://user-images.githubusercontent.com/37534529/190095464-8d29143d-2f32-47d7-94f9-25a9d699106e.png">

Perhaps move the page name above the description?

Also, it seems you've forgotten to sort emums alphabetically

penev92 committed:
Added documentation page title

penev92 commented:
Updated.

PunkPun reviewed (approved):
LGTM

penev92 commented:
I guess the question is whether @Mailaender's +1 still stands after the changes.

Mailaender commented:
> The shape of the beam. Accepts values Cylindrical or Flat.

Those descriptions are now redundant.

PunkPun commented:
Yes, all enum descriptions will need to be reevaluated, but perhaps in a follow PR

penev92 commented:
Yes, trait/weapon/sequence field descriptions would need to be revised to remove such listings, but I'd prefer to leave that for a follow-up PR. Also as suggested by @pchote I plan on adding `DescAttribute`s to enum types, which will then call for another pass over trait/weapon/sequence field descriptions to remove redundant explanations. I'd prefer if those two could happen in one pass.

penev92 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/206dbbd52eb5f5290c16623a4c1f5022fb534f8c...c3d38e09f232fa28b314165370e86e8861096a73)
